### PR TITLE
Enable Gemma3 MaxText-to-Huggingface Conversion

### PR DESCRIPTION
**Pull request by YixuanWang-99**
# Description

Enable Gemma3 models checkpoint conversion; from Maxtext to HuggingFace.  

- Added param mapping, shape mapping, model config of gemma3 models
- Added top token comparisons and kl divergence check
- Added more test prompts
- Minor fix about shell script path, max_logging usage, etc.

# Tests

Due to vocab_size difference between Maxtext model and Huggingface model, we can not [compare the weights directly](https://screenshot.googleplex.com/6eNZ8qaU5Kdc3MG). Instead, we compared the top tokens and tokens kl divergence. [Example](https://screenshot.googleplex.com/87TQWbfFrWgFu9L). 
We tested the Gemma3-4b model. 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/tests/hf_checkpoint_conversion_check.py [hunk 1]
```diff
@@ -70,6 +72,116 @@ def get_logits(inputs, model, golden_model):
...
+  kl_divs_per_token = F.kl_div(
+      input=model_log_probabilities, target=golden_probabilities, reduction="none", log_target=False
+  ).sum(
+      dim=-1
+  )  # Sum over the vocab dim to get a single KL value per token
+
+  max_kl_div = kl_divs_per_token.max()
+  print(f"\nMax KL divergence for a single token in the set: {max_kl_div.item():.6f}")
+
+  assert kl_div_value < atol, "KL divergence values exceed the tol"
```
Comment (hengtaoguo) [inline MaxText/tests/hf_checkpoint_conversion_check.py]: Can you do f"KL divergence values {kl_div_value} exceed the threshold {atol}"?

File: MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_hf.sh
Comment (hengtaoguo) [inline MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_hf.sh:1]: Do you plan to also add a bash scripts example for gemma3?

File: MaxText/tests/hf_checkpoint_conversion_check.py [hunk 2]
```diff
@@ -82,33 +194,47 @@ def main(argv: Sequence[str]) -> None:
...
+  inputs = tokenizer("Hello", return_tensors="pt")
+  model_logits, golden_model_logits = get_logits(inputs, model, golden_model)
+
+  tokens = get_top_k_tokens_scores(model_logits, tokenizer, k=10, description="converted model")
+  golden_tokens = get_top_k_tokens_scores(golden_model_logits, golden_tokenizer, k=10, description="golden model")
+  compare_top_tokens(converted_tokens=tokens, golden_tokens=golden_tokens)
 
-  # Run forward pass to get logits
-  logits, golden_logits = get_logits(inputs, model, golden_model)
+  check_kl_divergence(model_logits, golden_model_logits)
```
Comment (hengtaoguo) [inline MaxText/tests/hf_checkpoint_conversion_check.py]: We may prefer to compare the next token logits using the three prompts below ["I love to", "Today is a", "What is the"], instead of "Hello". Could you give it a try and see whether the results match?

File: MaxText/tests/hf_checkpoint_conversion_check.py [hunk 3]
```diff
@@ -70,6 +72,116 @@ def get_logits(inputs, model, golden_model):
...
+      "jaccard_similarity": jaccard_similarity,
+      "rank_agreement_percentage": rank_agreement,
+  }
+
+  print("\n--- Similarity Metrics of Top Tokens ---")
+  table = [[key, value] for key, value in metrics.items()]
+  print(tabulate(table, headers=["Metric", "Value"], tablefmt="orgtbl"))
+
+
+def check_kl_divergence(model_logits, golden_logits, atol=0.01):
```
Comment (hengtaoguo) [inline MaxText/tests/hf_checkpoint_conversion_check.py]: For this function's input atol value, can we make it an argument for this script, similar to [this](https://github.com/AI-Hypercomputer/maxtext/blob/ff6c8aca080eb55cde3e26ccbcdba5fbe0ebe8b9/MaxText/tests/forward_pass_logit_checker.py#L162)? You can set it default to 0.01 though.

The reason is that, in the future if you have other models with higher/lower kl_div, you can easily set higher/lower threshold for them specifically in example bash scripts and let them all pass (at least when the output texts make sense).

File: MaxText/tests/hf_checkpoint_conversion_check.py [hunk 4]
```diff
@@ -16,12 +16,14 @@
 
 from typing import Sequence
 import torch
+import torch.nn.functional as F
 from transformers import AutoTokenizer, AutoModelForCausalLM
 import os
-from absl import app  # Removed flags
+from absl import app
+from tabulate import tabulate
```
Comment (hengtaoguo) [inline MaxText/tests/hf_checkpoint_conversion_check.py:21]: Be aware that tabulate is not included in our [requirements](https://github.com/AI-Hypercomputer/maxtext/blob/main/requirements.txt). But since this file already imports torch and not run as unit tests, I think it is fine.

File: MaxText/tests/hf_checkpoint_conversion_check.py (hunk 1 above)
Comment (YixuanWang-99) [reply MaxText/tests/hf_checkpoint_conversion_check.py]: Fixed

File: MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_hf.sh
Comment (YixuanWang-99) [reply MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_hf.sh:1]: Added

File: MaxText/tests/hf_checkpoint_conversion_check.py (hunk 2 above)
Comment (YixuanWang-99) [reply MaxText/tests/hf_checkpoint_conversion_check.py]: Revised.

File: MaxText/tests/hf_checkpoint_conversion_check.py (hunk 4 above)
Comment (YixuanWang-99) [reply MaxText/tests/hf_checkpoint_conversion_check.py:21]: tabulate is used to make the output tables look better. We can remove it if requested

File: MaxText/tests/hf_checkpoint_conversion_check.py (hunk 3 above)
Comment (YixuanWang-99) [reply MaxText/tests/hf_checkpoint_conversion_check.py]: This is a good idea. Added atol arg

File: MaxText/tests/hf_checkpoint_conversion_check.py [hunk 5]
```diff
@@ -70,46 +71,181 @@ def get_logits(inputs, model, golden_model):
...
+  for i in range(k):
+    tok_id = topk_results.indices[i].item()
+    score = topk_results.values[i].item()
+    tok = tokenizer_instance.decode(tok_id)
+    collected_tokens.append({"id": int(tok_id), "token": tok.strip(), "score": float(score)})
+    tokens.append({"id": int(tok_id), "token": tok.strip(), "score": float(score)})
+
+  # Prepare data for tabulate: a list of lists
+  table_data = [[d["id"], d["token"], d["score"]] for d in collected_tokens]
+  print(tabulate(table_data, headers=["Token ID", "Token", "Score"], tablefmt="orgtbl"))
```
Comment (RissyRan) [inline MaxText/tests/hf_checkpoint_conversion_check.py]: Nit: I probably forgot to mention in previous PR. Could we use `max_logging.log` to align in the repo?

**Review by RissyRan — APPROVED** [review]
Just one minor comment. Thanks!

File: MaxText/tests/hf_checkpoint_conversion_check.py (hunk 5 above)
Comment (YixuanWang-99) [reply MaxText/tests/hf_checkpoint_conversion_check.py]: Revised. Thanks!

File: .github/workflows/UploadDockerImages.yml [hunk 6]
```diff
@@ -20,73 +20,153 @@
...
+          tags: |
+            gcr.io/tpu-prod-env-multipod/${{ matrix.image_name }}:${{ steps.date.outputs.image_date }}
+            gcr.io/tpu-prod-env-multipod/${{ matrix.image_name }}:latest
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          build-args: |
+            ${{ matrix.build_args }}
+            JAX_AI_IMAGE_BASEIMAGE=${{ matrix.base_image }}
+            COMMIT_HASH=${{ steps.vars.outputs.sha_short }}
+            DEVICE=gpu
```
[the remark this answers is not in the document]
Comment (YixuanWang-99) [reply .github/workflows/UploadDockerImages.yml]: No idea where this error comes from

File: .github/workflows/UploadDockerImages.yml [hunk 7]
```diff
@@ -20,73 +20,153 @@
...
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          build-args: |
+            ${{ matrix.build_args }}
+            JAX_AI_IMAGE_BASEIMAGE=${{ matrix.base_image }}
+            COMMIT_HASH=${{ steps.vars.outputs.sha_short }}
+            DEVICE=tpu
+
+  # Same as tpu-build step but mirrored for GPUs
+  build-gpu:
```
[the remark this answers is not in the document]
Comment (YixuanWang-99) [reply .github/workflows/UploadDockerImages.yml]: Don't know why this error pops
